test/TestNamespaceTracker.hs

```diff
@@ -127,62 +124,46 @@ collectAllResponsesUntil i = inner mempty
       | otherwise = inner bs
 
 fakeRelay ::
-    (Monad m, Show i, Show u) =>
-    Protocol (ClientEvent (AddrWithUser i u) [Om] x) Void (ServerEvent (AddrWithUser i u) [RoutableMessage]) Void m ()
+    (Monad m, Show c) =>
+    Protocol ((c, Request)) Void ((c, Response)) Void m ()
 fakeRelay = forever $ waitThen fwd undefined
   where
-    fwd (ClientConnect _ _) = return ()
-    fwd (ClientData i oms) = sendRev $ ServerData i $ map (mkRoutable i) oms
-    fwd (ClientDisconnect i) = sendRev $ ServerDisconnect i
-    mkRoutable i (Client, MsgSubscribe p) = RoutableMessage Nothing (MsgAssignType p root)
-    mkRoutable _ (o, m) = RoutableMessage (Just o) m
-
-alice = newAwu 42 "alice"
-alice' = newAwu 43 "alice"
-bob = newAwu 121 "bob"
-charlie = newAwu 7 "charlie"
-dave = newAwu 14 "dave"
-ethel = newAwu 96 "ethel"
+    fwd (ctx, Request mGets updates) = sendRev (ctx, Response (getResps mGets) [] updates)
+    getResps Nothing = []
+    getResps (Just ps) = Left . flip UMsgAssignType helloP <$> ps
+
+alice = "alice"
+alice' = "alice'"
+bob = "bob"
 
 helloP = [pathq|/hello|]
 
 testMessageToPreowned =
   let
-    owners = Map.singleton "" (newAwu 0 "relay itself")
+    owners = Map.singleton "hello" "relay itself"
     protocol = forTest <-> namespaceTrackerProtocol owners mempty <-> fakeRelay
     forTest = do
-      sendFwd $ ClientData alice [msg root Error]
-      sendFwd $ ClientDisconnect alice
-      resps <- collectAllResponsesUntil alice
-      lift $ assertOnlyKeysInMap [alice] resps
-      lift $ assertSingleError alice root ["forbidden", "client"] resps
-  in
-    runEffect protocol
-
-
-testSubscribeUnclaimed =
-  let
-    protocol = forTest <-> namespaceTrackerProtocol mempty mempty <-> fakeRelay
-    forTest = do
-      sendFwd $ ClientData alice [msg helloP Subscribe]
+      sendFwd $ ClientData alice $ Left $ UpdateBundle [] [Left $ UMsgAssignType helloP root]
       sendFwd $ ClientDisconnect alice
       resps <- collectAllResponsesUntil alice
       lift $ assertOnlyKeysInMap [alice] resps
-      lift $ assertSingleError alice helloP ["forbidden", "owner"] resps
+      lift $ assertSingleError alice helloP ["Client", "update"] resps
   in
     runEffect protocol
 
 testSubscribeAsOwner =
   let
-    protocol = forTest <-> namespaceTrackerProtocol mempty mempty <-> fakeRelay
+    -- Because the protocol is forward biased at the moment, we haven't dealt
```

Comment:
Is this true any more? Didn't you first commit in this PR change the bias in the relay to make it "left" or "reverse" biased?

Reply:
Ah, I meant to delete that.